core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.encryption;
+
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import javax.crypto.AEADBadTagException;
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+public class Ciphers {
+
+  public static class AesGcmEncryptor {
+    public static final int NONCE_LENGTH = 12;
+    public static final int GCM_TAG_LENGTH = 16;
+    public static final int GCM_TAG_LENGTH_BITS = 8 * GCM_TAG_LENGTH;
+
+    private final SecretKeySpec aesKey;
+    private final Cipher cipher;
+    private final SecureRandom randomGenerator;
+
+    AesGcmEncryptor(byte[] keyBytes) {
+      int keyLength = keyBytes.length;
+      if (!(keyLength == 16 || keyLength == 24 || keyLength == 32)) {
+        throw new IllegalArgumentException("Wrong key length " + keyLength);
+      }
+      this.aesKey = new SecretKeySpec(keyBytes, "AES");
+
+      try {
+        this.cipher = Cipher.getInstance("AES/GCM/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to create GCM cipher", e);
+      }
+
+      this.randomGenerator = new SecureRandom();
+    }
+
+    public byte[] encrypt(byte[] plainText) {
```

Comment:
It would be great to have another API to allow to input the length of the plainText. During the Parquet [PR](https://github.com/apache/parquet-mr/pull/928), I found sometimes we might need to reuse some array and that the data to be encrypted only take partially of the array.  If we don't provide that API, we are forced to dump the data to a new array with exactly the length of the data to be encrypted.

Reply:
agreed.